src/server/worker.h

```diff
@@ -88,9 +91,10 @@ class Worker : EventCallbackBase<Worker>, EvconnlistenerBase<Worker> {
   UniqueEvent timer_;
   std::thread::id tid_;
   std::vector<evconnlistener *> listen_events_;
-  std::mutex conns_mu_;
-  std::map<int, redis::Connection *> conns_;
-  std::map<int, redis::Connection *> monitor_conns_;
+  mutable std::mutex conns_mu_;  // refer to https://github.com/oneapi-src/oneTBB/issues/183, traverse and erase should
```

Comment:
Can we reference to the document of TBB? 🤔 An issue might be closed or fixed later so a official doc which point out the concurrent behavior here might be beeter

Reply:
Hi @mapleFU ,

Thanks very much for your review suggestions!

In fact I tried to get some official document after the TSAN issue emerged.
But It seems the tbb official document has no description on this behavior.

I finally found the problem in a [stackoverflow post](https://stackoverflow.com/questions/77673606/when-find-of-tbbconcurrent-hash-map-is-used-in-parallel-with-iteration-the).
This detail page was in intel's documentation site but now disappeared.

With the web archive, I got these two pages archive.
[Traversing concurrent_hash_map concurrently](https://web.archive.org/web/20200818130444/https://software.intel.com/content/www/us/en/develop/blogs/traversing-concurrent_hash_map-concurrently.html).
[concurrent_hash_map iterators](https://web.archive.org/web/20240429165704/https://spec.oneapi.io/versions/latest/elements/oneTBB/source/containers/concurrent_hash_map_cls/iterators.html)

The first archive has below sentences.
> Although it is not documented, you can traverse tbb::concurrent_hash_map concurrently if you are fine with the following restrictions and possible workarounds for them:
>
> You cannot traverse concurrently with an erase() operation. Use a mutex around erase and traversal code if you still need both running in different threads.
> - Some items can be missed due to concurrent rehashing. Try to reserve enough buckets before traversal using rehash() to prevent this or repeat the pass.
> - Some items can be visited twice or more times. Try to reserve enough buckets before traversal using rehash() to prevent this or put everything to a local serial map first to remove duplicates.
> - Access through iterators is not protected. Know what are you doing, e.g. use additional synchronization primitives like tbb::atomic to protect one member or acquire accessor to protect the rest.

In the second archive captured on 2024/04/29, there are some sentences on iteration.
> All member functions in this section can only be performed serially. The behavior is undefined in case of concurrent execution of these member functions with other (either concurrently safe) methods.

I don't know the exact reason of removing this page on current tbb documentation. 🤣
The latest official tbb release version and our dependency are same `oneapi-src/oneTBB v2021.13.0`.
It should not have any change on stable releases since the second archive published.
At least in my [GitHub actions](https://github.com/LindaSummer/kvrocks/actions/runs/10130746250/job/28017077204) this issue still exists.

Maybe we should create a new issue on tbb community for more official answers? 😄

Best Regards,
Edward

Reply:
https://github.com/oneapi-src/oneTBB/blob/master/include/oneapi/tbb/concurrent_hash_map.h#L1043-L1046

What about here?

Reply:
Besides, I'm still curious that what did the mutex protect after this patch. Since we use open-tbb now, the concurrent hash map would brings up much more cost ( on both memory usage and search time) than std::unordered_map, would a extra lock and synchronization break the potential performance speedup and causing the performance worse than before? Would you mind add `// guarded by` for the protected element?

Reply:
Hi @mapleFU ,

Thanks for your suggestions and help! ❤
I will add the tbb's code reference in comment.

> Would you mind add // guarded by for the protected element?

Of course, I will add comments for resources protected by mutex in next patches.

> would a extra lock and synchronization break the potential performance speedup and causing the performance worse than before?

Yes, during my development, I also concern about the performance of using both `tbb::concurrent_hash_map` and `std::mutex`. Maybe a simple read-write shared_mutex have a better performance and cleaner logic.

Best Regards,
Edward

Reply:
Hi @mapleFU ,
  Thanks for your suggestions!

  I have been inspired by your reference of [tbb source file](https://github.com/oneapi-src/oneTBB/blob/v2021.13.0/include/oneapi/tbb/concurrent_hash_map.h#L1033-L1051).
  I switched from the cpp style for loop to `tbb::parallel_for` and `tbb_parallel_reduce`.
  The `range()` method is commented as `Parallel algorithm support`.
  It works now and the mutex has been removed. 😊

  I read the source code of `tbb::parallel_recude` and `tbb::concurrent_hashmap::range()` and it seems that the partitioner will help [split the range](https://github.com/oneapi-src/oneTBB/blob/v2021.13.0/include/oneapi/tbb/partitioner.h#L273-L291) before execution.

Best Regards,
Edward